### PR TITLE
Test logging option passing

### PR DESCRIPTION
**Pull request by overlookmotel**
Please see issue https://github.com/sequelize/sequelize/issues/3834

### CODE REVIEW

**Comment by BridgeAR** [comment]
It's great that you take care of this! :) +1

**Comment by BridgeAR** [comment]
Just a thought: I'd say let's run the shims only if you pass a parameter for it. That way it's possible to run a single task to test the logging / transactions and run all other tasks without the shims.

**Comment by overlookmotel** [comment]
Thanks @BridgeAR. Please see the discussion under issue #3834.

**Comment by overlookmotel** [comment]
I've now done quite a few revisions on this.

Here's what it does:

* Shims all async Sequelize methods
* When a method is called from within the tests, the shim adds a `logging` function to `options`
* When called *within* Sequelize, the shim checks that `options.logging` has been passed to it correctly - and throws an error if not

It looks at hints made in comments in the code itself to work out which functions to shim, and where the `options` argument is to be found. These comments are of the form e.g. `// testhint options:none`.

I used this PR to to find all the bugs fixed in https://github.com/sequelize/sequelize/pull/3846. But, as discussed in https://github.com/sequelize/sequelize/issues/3834, the end game is to do the same kind of automated testing for `options.transaction` being passed correctly which is much more critical.

Now might be a good time for you guys to have a look at this PR and see if you're likely to be willing to merge it.

The code relating to hints would probably be best moved into a separate module, but for now I'm leaving it as is. Basically, I've done quite a lot of work on this, and don't want to do any more if you're not going to merge it! Or if discussion dictates I need to change approach...

Any thoughts please @mickhansen, @janmeier, @BridgeAR?

**Comment by janmeier** [comment]
Looks good - I think this is the way to go forward

https://www.npmjs.com/package/multilang-extract-comments Seems to support both multi and single line commenst well in my small tests, so it might be a candidate for identifying comments

**Comment by janmeier** [comment]
Great work @overlookmotel - I definitely think we should use this framework for transactions as well

**Comment by overlookmotel** [comment]
Thanks @janmeier!

I've now moved the code hints part into a module [hints](https://www.npmjs.com/package/hints), which tidies this PR a bit.

Ready for merge now, I think. Then I'll have a go at adapting it to check for passing of `options.transaction` too...

**Comment by janmeier** [comment]
@overlookmotel Might want to wait for @mickhansen to chime in as well :)

**Comment by mickhansen** [comment]
I still need to go through it in full, for some reason i'm a tad scared of shimming all the methods :D Not sure it's entirely rational though.

**Comment by overlookmotel** [comment]
@mickhansen Yes, I can see that it feels like quite a wide-ranging thing to do, so I totally understand your hesitance.

Well, see what you think... I wracked my brains and couldn't find any other way to ensure complete coverage for this issue. And I think it's absolutely critical that correct passing of `options.transaction` is tested comprehensively (which this PR is a step towards). So my logic was that it needs a heavy-duty approach to tackle a heavy-duty problem!

**Comment by overlookmotel** [comment]
Any further thoughts @mickhansen?

**Comment by mickhansen** [comment]
Sick with the flue at the moment @overlookmotel, promise to get back to you asap.

**Comment by overlookmotel** [comment]
@mickhansen Oh yuk. Get well soon...

**Comment by overlookmotel** [comment]
Hi @mickhansen. Just wondered if you'd had time to think any more about this? I'm keen to also implement the testing for correct passing of `transaction`.

**Comment by BridgeAR** [comment]
@overlookmotel couldn't you write the tests for passing transaction nevertheless this is merged or not? This works already helped a lot even if it's not merged to run every time the tests are run :) So there is a huge benefit by it no matter if it's in the code in the end or not.

File: test/supportShim.js [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+    if (original.__testShim) return;
+
+    if (original.__testShimmedTo) {
+      obj[name] = original.__testShimmedTo;
+    } else {
+      obj[name] = wrapper(original);
+      obj[name].__testShim = original;
+      original.__testShimmedTo = obj[name];
+    }
+  }
```
Comment (mickhansen) [inline test/supportShim.js]: Perhaps the `shimmer` package could be of help here instead?

**Comment by mickhansen** [comment]
Just took a quick looks and it looks alright, the code is a bit hard to follow but that might be me being sick.
Right now this is only called during a code coverage check correct?

**Comment by overlookmotel** [comment]
@mickhansen Actually the opposite. It is called on every test run *except* code coverage. As code coverage tests also shims methods, the two weren't interacting well.

My logic was that these tests should be run everywhere possible - otherwise, it might not catch a bug that is specific to a particular dialect (there were some bugs like this).

File: test/supportShim.js (hunk 1 above)
Comment (overlookmotel) [inline test/supportShim.js]: @mickhansen Yes, I used `shimmer` originally, but I needed to add some customizations (`__testShim` and `__testShimmedTo`), so in the end it was easier to remove `shimmer`.

**Comment by overlookmotel** [comment]
@mickhansen @janmeier Sorry to hassle, but have you had any further thoughts about this?

**Comment by mickhansen** [comment]
Looked at it again, it will probably take some time to become completely comfortable with it since i didn't author it.

But i think it's a great idea, let's get it merged as soon as tests pass.

**Comment by overlookmotel** [comment]
Tests are pending because I just rebased it on current master. If tests fail it means more bugs have crept into the Sequelize codebase!

I totally understand that you're wary - it makes me a bit squeemish too to be shimming all Sequelize's methods. But I couldn't see any other way to do it...

**Comment by overlookmotel** [comment]
@mickhansen Amazing! Thank you!
